codegen/CodeGenerator/src/main/java/AggregateProcessFunctionWriter.java

```diff
@@ -0,0 +1,106 @@
+
+    public AggregateProcessFunctionWriter(final AggregateProcessFunction aggregateProcessFunction) {
+        requireNonNull(aggregateProcessFunction);
+        this.aggregateProcessFunction = aggregateProcessFunction;
+        Class type = aggregateProcessFunction.getValueType();
+        aggregateType = type.equals(Type.getClass("date")) ? type.getName() : type.getSimpleName();
+        className = aggregateProcessFunction.getName();
+    }
+
+    @Override
+    public void generateCode(String filePath) throws IOException {
+        addImports();
+        addConstructorAndOpenClass();
+        addAggregateFunction();
+        addAdditionFunction();
+        addSubtractionFunction();
+        addInitStateFunction();
+        //class closing
+        writer.writeln_l("}");
+        Files.write(Paths.get(filePath + File.separator + className + ".scala"), writer.toString().getBytes());
+    }
+
+    private void addImports() {
+        writer.writeln("import org.hkust.RelationType.Payload");
+        writer.writeln("import org.apache.flink.api.common.state.ValueStateDescriptor");
+        writer.writeln("import org.apache.flink.api.common.typeinfo.{TypeHint, TypeInformation}");
+        writer.writeln("import org.hkust.BasedProcessFunctions.AggregateProcessFunction");
+    }
+
+    private void addConstructorAndOpenClass() {
+        List<AggregateProcessFunction.AggregateValue> aggregateValues = aggregateProcessFunction.getAggregateValues();
+        writer.writeln_r("class " + className + " extends AggregateProcessFunction[Any, " + aggregateType + "](\"" + className + "\", Array(), Array(),aggregateName = \""
+                + (aggregateValues.size() == 1 ? aggregateValues.get(0).getName() : "_multiple_")
+                + "\") {");
+    }
+
+    private void addAggregateFunction() {
+        writer.writeln_r("override def aggregate(value: Payload): " + aggregateType + " = {");
+        List<AggregateProcessFunction.AggregateValue> aggregateValues = aggregateProcessFunction.getAggregateValues();
+        aggregateValues.forEach(aggregateValue -> {
+            StringBuilder code;
+            if (aggregateValue.getType().equals("expression")) {
+                Expression expression = (Expression) aggregateValue.getValue();
+                List<Value> values = expression.getValues();
+                for (int i = 0; i < values.size(); i++) {
+                    code = new StringBuilder();
+                    Value value = values.get(i);
+                    if (!(value instanceof AttributeValue)) {
+                        throw new RuntimeException("Currently only attribute values supported for expression values");
+                    }
+                    AttributeValue attributeValue = (AttributeValue) value;
+                    String columnName = attributeValue.getName();
+                    Class type = RelationSchema.getColumnAttribute(columnName.toLowerCase()).getType();
+                    code.append("value(\"")
```

Comment:
I still think this part should be in a function for reuse, as it is used in multiple places.   Maybe you should have a look at RelationProcessFunctionWriter.java makeIsValidFunction(), abstract the same logic, and reuse those codes.

Reply:
umm making `makeIsValidFunction` abstract and inheriting it in the 2 Writer classes wouldn't be accurate as there is no `IsValidFunction` in `AggregateProcessFunction` but that inheritance structure will force me to to have a method called `makeIsValidFunction` in `AggregateProcessFunction`